### PR TITLE
#196 add `  hash  ` and `  eq  ` functionality to all classes

### PR DESCRIPTION
**Pull request by javier123454321**
Updated the __eq__() method to check for the keys. I only used the parameters that were currently being evaluated in the existing __eq__() method.

### CODE REVIEW

File: ladybug/analysisperiod.py [hunk 1]
```diff
@@ -463,3 +452,16 @@ def __repr__(self):
              self.end_time.month, self.end_time.day,
              self.st_time.hour, self.end_time.hour,
              self.timestep)
+
+    def __key(self):
+        return(self.st_time, self.end_time, self.timestep, self.is_leap_year)
```
Comment (chriswmackey) [inline ladybug/analysisperiod.py]: I think you need to `hash()` the datetime objects in order for the `__hash__` method to work correctly. We'd know for sure if you add a test like this:
```
a_period = AnalysisPeriod()
hash(a_period )
```
And you get an exception.

File: tests/analysisperiod_test.py [hunk 2]
```diff
@@ -180,3 +180,11 @@ def test_0_end_hour():
...
     assert ap.end_hour == 0
     assert len(ap) == 24
+
+def test_equality():
+    ap_one = AnalysisPeriod()
+    ap_two= AnalysisPeriod()
+    ap_three = AnalysisPeriod(st_month=2, end_month=6)
+
+    assert ap_one == ap_two
+    assert ap_one != ap_three
```
Comment (chriswmackey) [inline tests/analysisperiod_test.py]: This test should be a bit more robust. Here is a good template for you to follow when writing tests for these operators:
https://github.com/ladybug-tools/ladybug-geometry/blob/master/tests/pointvector3d_test.py#L34-L45

**Comment by chriswmackey** [comment]
Also, you will need to digitally sign the contributors license agreement (CLA) before we can merge.

**Comment by javier123454321** [comment]
Ok so It seems that the object is hashing fine when analyzing the keys of the analysisperiod object, and the new test cover that case. It works fine as is.

As far as the contributor agreement, I did the changes without an email attached to my git, therefore making the 3 original commits to come from an unknown id to github. I tried to ammend the commits with an email attached, but the new commits with the email that signed the CLA show as a different history. (this is getting to messy git territory that I am unfamiliar with) I tried to clean it as best I could.

**Comment by chriswmackey** [comment]
Thanks for adding the tests and I can see why it all passes now. I forgot that the ladybug datetime inherits from the native python datetime, which already has a `__hash__` function on it. 

All looks good, then. We can merge this in as soon as you sign the CLA.

**Comment by javier123454321** [comment]
Finally rebased the initial commits to change the author and put the right email in the history. This overwrote the original history, but is now all done, and the CLA is finally signed with the updated changes. I will continue with this task for another class using this framework as reference.

**Pull request merged**